lcls-twincat-motion/Library/Tests/FB_AtPositionState_Test.TcPOU

```diff
@@ -0,0 +1,521 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<TcPlcObject Version="1.1.0.1">
+  <POU Name="FB_AtPositionState_Test" Id="{cc4d1353-89e8-4ab0-b592-6491b8e8bd3e}" SpecialFunc="None">
+    <Declaration><![CDATA[FUNCTION_BLOCK FB_AtPositionState_Test EXTENDS TcUnit.FB_TestSuite
+(*
+    Test the following related helper functions:
+    - F_PosOverLowerBound
+    - F_PosUnderUpperBound
+    - F_PosWithinDelta
+    - F_AtPositionState
+*)
+VAR
+    // For the multi-cycle tests
+    stMotionStage: ST_MotionStage;
+    fbMotionStage: FB_MotionStage;
+    fbTestMove: FB_TestHelperSetAndMove;
+    stPositionStateInactive: ST_PositionState;
+    stPositionStateInvalid: ST_PositionState;
+    stPositionStateGood: ST_PositionState;
+    tonInactive: TON;
+    fbInternalGood: FB_PositionStateInternal;
+    fbInternalInvalid: FB_PositionStateInternal;
+    nTestCounter: UINT;
+    bOneTestDone: BOOL;
+END_VAR
+]]></Declaration>
+    <Implementation>
+      <ST><![CDATA[
+// Single cycle tests
+TestPosOverLowerBoundYes();
+TestPosOverLowerBoundNo();
+TestPosUnderUpperBoundYes();
+TestPosUnderUpperBoundNo();
+TestPosWithinDeltaTooLow();
+TestPosWithinDeltaTooHigh();
+TestPosWithinDeltaJustRight();
+
+// Multi cycle tests
+fbMotionStage(stMotionStage:=stMotionStage);
+stPositionStateInactive.sName := 'Inactive';
+stPositionStateInactive.fPosition := 30.0;
+stPositionStateInactive.fDelta := 1.0;
+stPositionStateInactive.bValid := TRUE;
+tonInactive(
+    IN:=TRUE,
+    PT:=T#1s,
+);
+TestAtPositionStateWithoutInternal();
+
+stPositionStateInvalid.sName := 'Invalid';
+stPositionStateInvalid.fPosition := 40.0;
+stPositionStateInvalid.fDelta := 1.0;
+stPositionStateInvalid.bValid := FALSE;
+fbInternalInvalid(
+    stMotionStage:=stMotionStage,
+    stPositionState:=stPositionStateInvalid,
+);
+TestAtPositionStateInvalid();
+
+stPositionStateGood.sName := 'Good';
+stPositionStateGood.fPosition := 50.0;
+stPositionStateGood.fDelta := 1.0;
+stPositionStateGood.bValid := TRUE;
+fbInternalGood(
+    stMotionStage:=stMotionStage,
+    stPositionState:=stPositionStateGood,
+);
+TestAtPositionStateTooLow();
+TestAtPositionStateTooHigh();
+TestAtPositionStateJustRight();
+TestAtPositionStateTweak();
+TestAtPositionStateLeave();
+
+IF bOneTestDone THEN
+    bOneTestDone := FALSE;
+    nTestCounter := nTestCounter + 1;
+    fbTestMove(
+        stMotionStage:=stMotionStage,
+        bExecute:=FALSE,
+    );
+END_IF
+
+]]></ST>
+    </Implementation>
+    <Method Name="TestAtPositionStateInvalid" Id="{cd076af2-9cac-4a9c-b69d-20d2d211c94e}">
+      <Declaration><![CDATA[METHOD PRIVATE TestAtPositionStateInvalid
+(*
+    If a position state is invalid, it is never the right state.
+*)
+VAR
+
+END_VAR
+]]></Declaration>
+      <Implementation>
+        <ST><![CDATA[
+TEST('TestAtPositionStateInvalid');
```

Comment:
Naively I would have thought the `TEST()` would go after the `nTestCounter` "mutex" (so to speak)

I guess the intent then is to indicate to TcUnit that there's all these tests pending, even if they haven't completed yet?

Reply:
This is part of the ugly nuance of setting up multiple cycle tests.

If `TcUnit` starts a plc cycle and there are no remaining unfinished `TEST` calls, it ends execution and displays the results of the test suite. So, you have to be careful not to end the test suite early without skipping any of your tests.

One way to work around this would be to set the new `TEST` call during the same cycle that the previous one ended, but now you have test-ordering-dependent behavior and you have to be super careful about it.

It's simplest to have every test declare that it exists in the first cycle, so `TcUnit` knows that it should still wait for their results.

Reply:
(One other alternative: each test could have been its own so-called "test suite", but I found it helpful to logically group things by which FB they were testing)